File: extension/lib/alias_manager.js
// 세션별 alias 매핑 관리 — alias_manager.py 의 JS 포팅.
// 같은 세션(=탭) 내 같은 entity 텍스트는 같은 alias 를 받는다.
// 정규식 토큰은 더미 템플릿 + 등장 순번을 부여한다.

import { ALIAS_POOLS, REGEX_DUMMY } from "./aliases.js";

const REGEX_LABELS = new Set(["PHONE", "EMAIL", "RRN", "CARD", "ACCOUNT", "IP", "API_KEY"]);
const NER_LABELS = new Set(["PERSON", "ORG", "LOCATION", "PROJ_N"]);

class SessionState {
  constructor() {
    this.mapping = new Map();      // "label original" -> alias
    this.nerCursor = new Map();    // label -> 다음 풀 index
    this.regexCursor = new Map();  // label -> 다음 순번(1부터)
  }
}

export class AliasManager {
  constructor() {
    this._sessions = new Map(); // sessionId -> SessionState
  }

  _session(sessionId) {
    let s = this._sessions.get(sessionId);
    if (!s) {
      s = new SessionState();
      this._sessions.set(sessionId, s);
    }
    return s;
  }

  /**
   * (label, original) → alias. 같은 세션에서 재등장하면 같은 alias 반환.
   */
  getAlias(sessionId, label, original) {
    const state = this._session(sessionId);
    const key = `${label} ${original}`;
    if (state.mapping.has(key)) return state.mapping.get(key);

    let alias;
    if (REGEX_LABELS.has(label)) {
      const idx = state.regexCursor.get(label) ?? 1;
      alias = REGEX_DUMMY[label](idx);
      state.regexCursor.set(label, idx + 1);
    } else if (NER_LABELS.has(label)) {
      const pool = ALIAS_POOLS[label];
      let cursor = state.nerCursor.get(label) ?? 0;
      // 가명이 원문의 부분문자열(또는 그 반대)이면 복원 시 무한 치환 위험
      // (예: 원문 "서울시 강남구" ⊃ 가명 "서울시") → 충돌 없는 다음 후보 선택.
      let tries = 0;
      do {
        alias = pool[cursor % pool.length];
        cursor += 1;
        tries += 1;
      } while (tries < pool.length && (original.includes(alias) || alias.includes(original)));
      state.nerCursor.set(label, cursor);
    } else {
      alias = `[${label}]`; // 알 수 없는 라벨 fallback
    }

    state.mapping.set(key, alias);
    return alias;
  }

  /**
   * 세션의 모든 (alias, original) 쌍 반환 (unmask 용).
   * key 는 "LABEL original" 형식이라 첫 공백 이후를 original 로 본다.
   */
  getPairs(sessionId) {
    const s = this._sessions.get(sessionId);
    if (!s) return [];
    const pairs = [];
    for (const [key, alias] of s.mapping) {
      const original = key.slice(key.indexOf(" ") + 1);
      if (alias !== original) pairs.push({ alias, original });
    }
    return pairs;
  }

  clearSession(sessionId) {
    if (sessionId === "*") {
      const had = this._sessions.size > 0;
      this._sessions.clear();
      return had;
    }
    return this._sessions.delete(sessionId);
  }

  stats() {
    let total = 0;
    for (const s of this._sessions.values()) total += s.mapping.size;
    return { activeSessions: this._sessions.size, totalMappings: total };
  }
}

File: server/alias_manager.py
"""세션별 alias 매핑 관리.

같은 세션 내 같은 entity 텍스트는 같은 alias 받도록 보장.
정규식 토큰(PHONE 등)은 더미 값 + 등장 순서 번호 부여.
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Dict, Optional

import yaml

REPO_ROOT = Path(__file__).resolve().parents[1]
ALIASES_PATH = REPO_ROOT / "configs" / "aliases.yaml"

# NER 라벨 → aliases.yaml 키 매핑
NER_KEY_MAP = {
    "PERSON": "person",
    "ORG": "org",
    "LOCATION": "location",
    "PROJ_N": "proj_n",
}

# 정규식 토큰
REGEX_LABELS = ("PHONE", "EMAIL", "RRN", "CARD", "ACCOUNT", "IP", "API_KEY",
                "PASSPORT", "DRIVER_LICENSE", "BIZ_NUM")


@dataclass
class SessionState:
    """단일 세션의 entity → alias 매핑."""
    # (label, original) → alias
    mapping: Dict[tuple, str] = field(default_factory=dict)
    # 라벨별 다음 풀 index
    ner_cursor: Dict[str, int] = field(default_factory=dict)
    # 정규식 라벨별 다음 순번 (1부터)
    regex_cursor: Dict[str, int] = field(default_factory=lambda: {l: 1 for l in REGEX_LABELS})
    last_access: float = field(default_factory=time.time)


class AliasManager:
    """세션별 alias 매핑 매니저 (thread-safe).

    동작:
      get_alias(session_id, label, original_text) -> alias
        - 같은 (label, original) 이 같은 세션에서 다시 들어오면 같은 alias
        - 새 entity 면 풀에서 다음 alias 할당 (순환)
        - 정규식 라벨은 더미 템플릿에 번호를 채워 반환
    """

    def __init__(self, aliases_path: Path = ALIASES_PATH):
        with open(aliases_path, encoding="utf-8") as f:
            cfg = yaml.safe_load(f)
        self._pools: Dict[str, list] = {
            "PERSON": cfg["person"],
            "ORG": cfg["org"],
            "LOCATION": cfg["location"],
            "PROJ_N": cfg["proj_n"],
        }
        self._regex_dummies: Dict[str, str] = cfg["regex_dummies"]
        self._sessions: Dict[str, SessionState] = {}
        self._lock = threading.Lock()

    def get_alias(self, session_id: str, label: str, original: str) -> str:
        key = (label, original)
        with self._lock:
            state = self._sessions.setdefault(session_id, SessionState())
            state.last_access = time.time()
            if key in state.mapping:
                return state.mapping[key]

            if label in REGEX_LABELS:
                template = self._regex_dummies[label]
                idx = state.regex_cursor[label]
                alias = template.format(i=idx)
                state.regex_cursor[label] = idx + 1
            elif label in NER_KEY_MAP:
                pool = self._pools[label]
                cursor = state.ner_cursor.get(label, 0)
                alias = pool[cursor % len(pool)]
                state.ner_cursor[label] = cursor + 1
            else:
                # 알 수 없는 라벨은 placeholder 로 fallback
                alias = f"[{label}]"

            state.mapping[key] = alias
            return alias

    def clear_session(self, session_id: str) -> bool:
        with self._lock:
            return self._sessions.pop(session_id, None) is not None

    def gc(self, max_age_seconds: float = 3600.0) -> int:
        """오래된 세션 garbage collection (1시간 이상 미사용)."""
        now = time.time()
        removed = 0
        with self._lock:
            for sid in list(self._sessions.keys()):
                if now - self._sessions[sid].last_access > max_age_seconds:
                    del self._sessions[sid]
                    removed += 1
        return removed

    def stats(self) -> dict:
        with self._lock:
            return {
                "active_sessions": len(self._sessions),
                "total_mappings": sum(len(s.mapping) for s in self._sessions.values()),
            }
